libraries/lib-numeric-formats/NumericConverterRegistry.h

```diff
@@ -0,0 +1,117 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file NumericConverterRegistry.h
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "Registry.h"
+#include "NumericConverterType.h"
+
+struct NumericConverterFormatter;
+
+struct NUMERIC_FORMATS_API NumericConverterConfig final
+{
+   double sampleRate { 44100.0 };
+
+   double tempo { 120.0 };
+   int upperTimeSignature { 4 };
+   int lowerTimeSignature { 4 };
+};
+
+using NumericConverterFormatterFactory =
+   std::function<std::unique_ptr<NumericConverterFormatter>(
+      const NumericConverterConfig&)>;
+
+struct NUMERIC_FORMATS_API NumericConverterRegistryGroup :
+    public Registry::InlineGroupItem<NumericConverterRegistryGroup>
```

Comment:
The template argument is not for CRTP.  It's to specify the type of argument supplied to Computed registry items, which you don't need in this registry.

So just use empty angle brackets to default the argument: `<>`

Reply:
This won't link due to the symbols duplication. This is the reason the Snap registry uses some type as well, but you have missed it probably because it is called Visitor as well. The `template< typename VisitorType = ComputedItem::DefaultVisitor ` is really confusing, it took me a lot of time to figure out that VisitorType is not related to the Visitor, rather it is a factory.

 I can use some random tag type here, if you think this would improve the code. 

TBH, I think Registries need quite a lot of work to be done.

Reply:
> This won't link due to the symbols duplication. This is the reason the Snap registry uses some type as well, but you have missed it probably because it is called Visitor as well. The `template< typename VisitorType = ComputedItem::DefaultVisitor ` is really confusing, it took me a lot of time to figure out that VisitorType is not related to the Visitor, rather it is a factory.
> 
> I can use some random tag type here, if you think this would improve the code.
> 
> TBH, I think Registries need quite a lot of work to be done.

Explain more some time.

I hope you recognize the good purpose they serve.  If you mean it should be easier to define new registries to serve that purpose, then I agree and I have some ideas.

Reply:
For one - I don't think `InlineGroupItem` should not be parametrized with a factory which is called visitor :-) 

In the end, you can only have single `dllexport`ed subclass using the default parametrization, otherwise Audacity won't link due to the symbol conflicts between libraries. So I had to use some kind of tagging to overcome this issue. Somehow you have missed `struct SnapRegistryGroup : public Registry::InlineGroupItem<SnapRegistryVisitor>` which is wrong too, because InlineGroupItem expects a factory, not a visitor :-)

Reply:
> For one - I don't think `InlineGroupItem` should not be parametrized with a factory which is called visitor :-)
> 
> In the end, you can only have single `dllexport`ed subclass using the default parametrization, otherwise Audacity won't link due to the symbol conflicts between libraries. So I had to use some kind of tagging to overcome this issue. Somehow you have missed `struct SnapRegistryGroup : public Registry::InlineGroupItem<SnapRegistryVisitor>` which is wrong too, because InlineGroupItem expects a factory, not a visitor :-)

No, it is really parametrized by a Visitor type, which can also act as the context expected by the ComputedItems.  Read this closer:
`void AppendOne( const ComputedItem::Factory<VisitorType> &factory )`
The VisitorType is a template parameter of a "factory" type alias but is not itself a factory type.  It's the input to the factory.

Example from PluginMenus:
```
      Section( "RepeatLast",
         // Delayed evaluation:
         [](AudacityProject &project)
         {
            const auto &lastEffect = MenuManager::Get(project).mLastEffect;
            TranslatableString buildMenuLabel;
            if (!lastEffect.empty())
               buildMenuLabel = XO("Repeat %s")
                  .Format( EffectManager::Get().GetCommandName(lastEffect) );
            else
               buildMenuLabel = XO("Repeat Last Effect");

            return Command( wxT("RepeatLastEffect"), buildMenuLabel,
               OnRepeatLastEffect,
               AudioIONotBusyFlag() | TimeSelectedFlag() |
                  WaveTracksSelectedFlag() | HasLastEffectFlag(),
               wxT("Ctrl+R") );
         }
      ),
```

Change the argument of the lambda to AudacityProject* (and project to *project inside) and it properly fails to compile the call to AppendOne.

The class `ToolbarMenuVisitor` (bad name, I will change it in the PR for lib-menus) implicitly converts to AudacityProject & and is also the template argument of the group items used in the menu registry.

Reply:
I can see how ToolbarMenuVisitor is a visiting factory, this doesn't help establishing that relationship when one tries to figure out how to use the Registry :-) Nor does it help with the linking problems or writing simple registries. For both registries introduced in this PR I had to hack my way through.

I understand what you have tried to achieve and I find this solution to be very elegant, but I have a strong opinion that this should be not be done by `InlineGroupItem`.  Instead, some well documented `ComputableGroupItem` can be used (naming is my weakness, yes).

Reply:
Some details may be revisited, but what I wanted was the convenient variadic calls to `Items` or `Section` or `Menu`, which usually accept `Command` but can sometimes just take a lambda in-line for the computed item.  The lambda must accept the correct argument type for that registry.  What that type is -- had to be given somehow as a template argument.

Reply:
Better comments, maybe newer C++17 template tricks for more convenience -- all of this is needed.  I need to look over Registry.* with fresh eyes after having first done it some years ago, and the pull request for registry-changes has been waiting as a draft for a long time, still needing its self review and cleanup as non-draft.

Then also I hope you will learn Registry inside and out -- read all of Registry.cpp and understand how this pull request makes some bug fixes and improvements.

Reply:
Please stay tuned for these improvements!